README.md
# express-session-id

[![Generated with nod](https://img.shields.io/badge/generator-nod-2196F3.svg?style=flat-square)](https://github.com/diegohaz/nod)
[![NPM version](https://img.shields.io/npm/v/express-session-id.svg?style=flat-square)](https://npmjs.org/package/express-session-id)
[![Build Status](https://travis-ci.org/dreisel/express-session-id.svg?branch=master)](https://travis-ci.org/dreisel/express-session-id) [![Coverage Status](https://img.shields.io/codecov/c/github/dreisel/express-session-id/master.svg?style=flat-square)](https://codecov.io/gh/dreisel/express-session-id/branch/master)

An express middleware for session ID.
Mainly Used for analytics session tracking.

## Install

    $ npm install --save express-session-id

## Usage

```js
import sessionId from 'express-session-id';
import express from 'express';
import cookieParser from 'cookie-parser';

const app = express();
app.use(cookieParser('my secret key'));
app.use(sessionId({
  idleTime: 10 * 1000 * 60, // 10 minutes
  cookie: {
    signed: true
  }
}))

app.get('/', (req, res) => {
  console.log(req.sessionID);
  /*
    Rest of the code
  */
});
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [sessionId](#sessionid)

### sessionId

This function return a session id middleware.
The session id is placed at req.sessionID

**Parameters**

-   `options` **[object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)**  (optional, default `{}`)
    -   `options.idleTime` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** max idle time before destoying a session. env: SESSION_ID_IDLE_TIME (optional, default `(30*1000*60)`)
    -   `options.name` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Name of the cookie. env: SESSION_ID_NAME (optional, default `s_id`)
    -   `options.genId` **[function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)** function to generate id's. see [uuid/v4](https://www.npmjs.com/package/uuid) (optional, default `uuid/v4`)
    -   `options.cookie` **[object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** config to pass to cookies. see [cookie](https://www.npmjs.com/package/cookie) (optional, default `{}`)

## License

MIT © [Daniel Reisel](https://github.com/dreisel)


CHANGELOG.md
<a name="2.0.4"></a>
## [2.0.4](https://github.com/dreisel/express-session-id/compare/v2.0.3...v2.0.4) (2020-03-31)



<a name="2.0.3"></a>
## [2.0.3](https://github.com/dreisel/express-session-id/compare/v2.0.2...v2.0.3) (2020-03-30)



<a name="2.0.2"></a>
## [2.0.2](https://github.com/dreisel/express-session-id/compare/v2.0.1...v2.0.2) (2020-03-30)



<a name="2.0.1"></a>
## [2.0.1](https://github.com/dreisel/express-session-id/compare/v2.0.0...v2.0.1) (2019-01-26)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/dreisel/express-session-id/compare/v1.0.2...v2.0.0) (2019-01-26)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/dreisel/express-session-id/compare/v1.0.1...v1.0.2) (2018-08-06)



<a name="1.0.1"></a>
## 1.0.1 (2018-08-06)


### Bug Fixes

* **build:** add support to require ([9c66da2](https://github.com/dreisel/express-session-id/commit/9c66da2))
* **req:** add sessionID to request ([ff69775](https://github.com/dreisel/express-session-id/commit/ff69775))



<a name="1.0.0"></a>
# 1.0.0 (2018-08-05)
* Initial









test/index.test.js
const express = require('express');
const cookieParser = require('cookie-parser');
const cookie = require('cookie');
const request = require('supertest');
const _ = require('lodash');
const uuid = require('uuid/v4');
const sessionId = require('../src');

const genIdMock = jest.fn();
const idleTime = 5 * 1000 * 60;
genIdMock.mockImplementation(uuid);

describe('config', () => {
  test('should throw invalid session id generator', () => {
    const badConfig = {
      genId: 1,
    };
    expect(() => sessionId(badConfig)).toThrow('genId must be a function');
  });
});
test('Cant run without cookie parser', async () => {
  const app = express();
  app.use(sessionId());
  const resp = await request(app).get('/');
  expect(resp.status).toEqual(500);
});
describe('sessionId', () => {
  let app;
  beforeEach(() => {
    app = express();
    app.use(cookieParser());
    app.use(
      sessionId({
        name: 'session_id',
        genId: genIdMock,
        idleTime,
      }),
    );
    app.get('/', (req, res) => {
      res.json({ status: 'success' });
    });
  });
  test('generates a new session', async () => {
    const response = await request(app).get('/');
    const sessionCookie = _.get(response, 'headers[set-cookie]', [])
      .map(cookie.parse)
      .find(x => x.session_id);
    expect(genIdMock).toHaveBeenCalledTimes(1);
    expect(sessionCookie).toBeDefined();
  });
  test('touch an existing session', async () => {
    const response = await request(app)
      .get('/')
      .set('Cookie', ['session_id=testing123']);
    const sessionCookie = _.get(response, 'headers[set-cookie]', [])
      .map(cookie.parse)
      .find(x => x.session_id);
    expect(Number.parseInt(sessionCookie['Max-Age'], 10) * 1000).toBe(idleTime);
    expect(sessionCookie.session_id).toBe('testing123');
  });
  test('test signed cookies', async () => {
    app = express();
    app.use(cookieParser('secret_key!'));
    app.use(
      sessionId({
        cookie: {
          signed: true,
        },
      }),
    );
    app.get('/session', (req, res) => {
      res.json({ sessionID: req.sessionID });
    });

    let response = await request(app).get('/session');
    const cookies = [null, null];
    const sessionIDs = [null, null];
    cookies[0] = _.get(response, 'headers[set-cookie]', []).find(x => x.substring(0, 4) === 's_id');
    sessionIDs[0] = _.get(response, 'body.sessionID');
    response = await request(app)
      .get('/session')
      .set('Cookie', [cookies[0]]);
    cookies[1] = _.get(response, 'headers[set-cookie]', []).find(x => x.substring(0, 4) === 's_id');
    sessionIDs[1] = _.get(response, 'body.sessionID');
    expect(sessionIDs[0]).toBeDefined();
    expect(sessionIDs[1]).toBeDefined();
    expect(sessionIDs[0]).toBe(sessionIDs[1]);
  });
});


src/index.js
import merge from 'lodash.merge';
import baseConfig from './config';

/**
 * This function return a session id middleware.
 * The session id is placed at req.sessionID
 * @param {object}   options
 * @param {number}   [options.idleTime=(30 * 1000 * 60)] - max idle time before destroying a session. env: SESSION_ID_IDLE_TIME
 * @param {string}   [options.name=s_id] - Name of the cookie. env: SESSION_ID_NAME
 * @param {function} [options.genId=uuid/v4] - function to generate id's. see {@link https://www.npmjs.com/package/uuid|uuid/v4}
 * @param {object}   [options.cookie={}] - config to pass to cookies. see {@link https://www.npmjs.com/package/cookie|cookie}
 */
function sessionId(options = {}) {
  const config = merge({}, baseConfig, options);
  const cookieName = config.name;
  if (typeof config.genId !== 'function') {
    throw new Error('genId must be a function');
  }
  return (req, res, next) => {
    if (typeof req.cookies === 'undefined') {
      return next(Error('cookie parser must be set before sessionId'));
    }
    let sessionCookie = req.cookies[cookieName] || req.signedCookies[cookieName];
    if (!sessionCookie) {
      sessionCookie = config.genId();
    }
    res.cookie(cookieName, sessionCookie, { ...config.cookie, maxAge: config.idleTime });
    req.sessionID = sessionCookie;
    return next();
  };
}

module.exports = sessionId;
export default sessionId;
